### ISSUE TITLE
Receving "double-spend" nofitications 

### ISSUE THREAD

**Issue by BullishNode** [mempool/mempool#4272]
### Description

We are using the WebSocket API to receive notifications for inbound transactions related to an address using `track-address`

What we want is to be notified if an unconfirmed transaction sent to that address has been "cancelled" via RBF or double-spent using any other method, i.e. that same inputs were used in a subsequent transaction.

### Proposed solution

In Cyphernode, we track both addresses as well as txids.

One possible way to do this would be that once we receive a transaction notification regarding an address we are tracking, we could track that specific txid and get notified if there is another transaction that is spending the same inputs as a the transaction we are tracking.

Another possible way would be to stick to tracking the address, and get notified if there is any transaction attempting to spend any of the inputs of any unconfirmed transaction of which the address we are tracking is an output of.

Note that for our use-case, we don't really care about tracking double spends concerning an address whose utxo(s) are inputs to a transaction. 

#### Additional info

The goal is to get this integrated in Cyphernode and serve as a real-time double-spend detection tool for businesses receiving payments, presumably Bitcoin payment processors. Receiving a real-time notification of double-spend can massively reduce the potential damage of a double-spend.

Note that in no way shape or form is this tool designed to make "0-conf" transactions safer (they will never be safe). However, a Bitcoin payment processor will usually "guarantee" the exchange rate to the merchant based on the 0-conf transaction. There is no way around this: if the merchant wants to receive $100 for a payment, the payment processor must be able to tell the merchant "if you collect 0.01 BTC I will credit your account with $100". The payment processor will usually hedge the fiat value of that incoming payment, which incurs a cost. If that payment is double-spent, even if the merchant did not send the merchandise to his client, the payment processor neverthelss is hedging the fiat value of the payment. If the payment processor is immediately notified of a double-spend (or RBF-cancel) he can immediately stop hedging the fiat value of that payment. The sooner he is able to do this, the less money he will lose on that hedging operation.

**Comment by mononaut** [comment]
We actually just merged a feature in #4209 which might be sufficient for this use-case:

The 'track-address' WebSocket subscription now sends notifications of related transactions removed from the mempool under the key `'address-removed-transactions'`:

https://github.com/mempool/mempool/blob/d0696628b25dc060b6adcc4dd379555f3900f0dd/backend/src/api/websocket-handler.ts#L529-L542

You could use this to identify double-spends/cancellations by filtering for removed funding transactions in `'address-removed-transactions'` without a corresponding replacement spending the same inputs in `'address-transactions'`.

Note that if you're running your own instance, `'address-removed-transactions'` will include transactions dropped for non-RBF reasons such as `maxmempool` or `mempoolexpiry` settings.

---

The `track-tx` subscription also sends an `rbfTransaction` event when a transaction is replaced for the first time, containing the txid of the new version, as well as more verbose `rbfInfo` events.

https://github.com/mempool/mempool/blob/d0696628b25dc060b6adcc4dd379555f3900f0dd/backend/src/api/websocket-handler.ts#L572-L590

You could use this to detect simple double-spends/cancellations by checking if the replacement still includes your expected output.

However, this may be less robust than the address subscription method, since it won't catch e.g. the case where a transaction is 'cancelled' by replacing an unconfirmed parent.

---

These features aren't officially documented and stabilized yet, but I'm sure they could be if there's a clear use-case beyond the Mempool frontend.

**Comment by BullishNode** [comment]
Fantastic thanks @mononaut will see if we can achieve our goal with the info you just gave!

**Comment by BullishNode** [comment]
@mononaut if our goal is only to get notified of an RBF for a given txid, we don't need to subscribe to track-rbf right? This would be for an explorer that wants to view the entiry rbf history?

**Comment by mononaut** [comment]
correct, `track-rbf` subscribes to *all* recent replacements, and is intended for something like the [mempool.space/rbf](https://mempool.space/rbf) page.

to get notified of an RBF for a specific transaction, it would be best to subscribe to `track-tx` for that txid and then listen for `rbfTransaction` events.
